File: DMBAlmanacProjectFolder/dmb-almanac-svelte/scraper/src/scrapers/liberation.ts
import { chromium, Page } from "playwright";
import * as cheerio from "cheerio";
import { writeFileSync } from "fs";
import { join, dirname } from "path";
import { delay } from "../utils/rate-limit.js";
import { cacheHtml, getCachedHtml } from "../utils/cache.js";
import { normalizeWhitespace, parseDate } from "../utils/helpers.js";

const BASE_URL = "https://www.dmbalmanac.com";
const OUTPUT_DIR = join(dirname(import.meta.url).replace("file://", ""), "../../output");

export interface ScrapedLiberationEntry {
  songId: string;
  songTitle: string;
  lastPlayedDate: string; // ISO format YYYY-MM-DD
  lastPlayedShowId: string;
  daysSince: number;
  showsSince: number;
  notes?: string;
  configuration: "full_band" | "dave_tim" | "dave_solo";
  isLiberated: boolean;
  liberatedDate?: string;
  liberatedShowId?: string;
}

export interface LiberationOutput {
  scrapedAt: string;
  source: string;
  totalItems: number;
  entries: ScrapedLiberationEntry[];
}

// Parse the Liberation List page
async function parseLiberationPage(page: Page): Promise<ScrapedLiberationEntry[]> {
  console.log("Fetching Liberation List...");

  const url = `${BASE_URL}/Liberation.aspx`;

  try {
    // Check cache first
    let html = getCachedHtml(url);

    if (!html) {
      await page.goto(url, { waitUntil: "networkidle", timeout: 30000 });
      html = await page.content();
      cacheHtml(url, html);
    }

    const $ = cheerio.load(html);
    const entries: ScrapedLiberationEntry[] = [];

    // Find the liberation list table
    const liberationTable = $("table.liberation-list");

    if (liberationTable.length === 0) {
      console.warn("Liberation list table not found");
      return [];
    }

    // Parse each row in the table body
    liberationTable.find("tbody tr").each((_, row) => {
      const $row = $(row);

      // Check if this is a liberated song (has special background color)
      const isLiberated = $row.attr("style")?.includes("background-color") || false;

      // Extract song title and ID
      const songLink = $row.find("td.rowcell.lj a[href*='summary.aspx']").first();
      if (songLink.length === 0) return;

      const songTitle = normalizeWhitespace(songLink.text());
      const songHref = songLink.attr("href") || "";
      const songIdMatch = songHref.match(/sid=(\d+)/);
      if (!songIdMatch) return;
      const songId = songIdMatch[1];

      // Extract last played date and show ID
      const lastPlayedLink = $row.find("td.rowcell.cj a[href*='TourShowSet']").first();
      if (lastPlayedLink.length === 0) return;

      const lastPlayedDateText = normalizeWhitespace(lastPlayedLink.text());
      const lastPlayedHref = lastPlayedLink.attr("href") || "";
      const showIdMatch = lastPlayedHref.match(/id=(\d+)/);
      if (!showIdMatch) return;
      const lastPlayedShowId = showIdMatch[1];

      // Parse date from format like "1.26.93" to "1993-01-26"
      let lastPlayedDate = "";
      const dateMatch = lastPlayedDateText.match(/(\d{1,2})\.(\d{1,2})\.(\d{2,4})/);
      if (dateMatch) {
        let [, month, day, year] = dateMatch;
        // Handle 2-digit years
        if (year.length === 2) {
          const yearNum = parseInt(year, 10);
          year = yearNum > 50 ? `19${year}` : `20${year}`;
        }
        lastPlayedDate = `${year}-${month.padStart(2, "0")}-${day.padStart(2, "0")}`;
      }

      // Extract days since and shows since
      const daysSinceCell = $row.find("td.rowcell.cj.d-none.d-sm-table-cell").eq(0);
      const showsSinceCell = $row.find("td.rowcell.cj.d-none.d-sm-table-cell").eq(1);

      const daysSince = parseInt(normalizeWhitespace(daysSinceCell.text()), 10) || 0;
      const showsSince = parseInt(normalizeWhitespace(showsSinceCell.text()), 10) || 0;

      // Extract notes
      const notesCell = $row.find("td.endcell.lj");
      let notes: string | undefined = normalizeWhitespace(notesCell.text());

      // Parse liberation info from notes if present
      let liberatedDate: string | undefined;
      let liberatedShowId: string | undefined;

      if (isLiberated) {
        const liberatedLink = notesCell.find("a[href*='TourShowSet']").last();
        if (liberatedLink.length > 0) {
          const liberatedHref = liberatedLink.attr("href") || "";
          const liberatedShowMatch = liberatedHref.match(/id=(\d+)/);
          if (liberatedShowMatch) {
            liberatedShowId = liberatedShowMatch[1];
          }

          const liberatedText = normalizeWhitespace(liberatedLink.text());
          const liberatedDateMatch = liberatedText.match(/(\d{1,2})\.(\d{1,2})\.(\d{2,4})/);
          if (liberatedDateMatch) {
            let [, month, day, year] = liberatedDateMatch;
            if (year.length === 2) {
              const yearNum = parseInt(year, 10);
              year = yearNum > 50 ? `19${year}` : `20${year}`;
            }
            liberatedDate = `${year}-${month.padStart(2, "0")}-${day.padStart(2, "0")}`;
          }
        }
      }

      // Determine configuration based on notes
      let configuration: "full_band" | "dave_tim" | "dave_solo" = "full_band";

      const notesLower = notes.toLowerCase();
      if (
        notesLower.includes("dave & tim") ||
        notesLower.includes("dave and tim")
      ) {
        configuration = "dave_tim";
      } else if (
        notesLower.includes("dave solo") ||
        notesLower.includes("played by dave solo")
      ) {
        configuration = "dave_solo";
      } else if (
        notesLower.includes("never played without") ||
        notesLower.includes("never fully played")
      ) {
        // Songs that can't be played without certain members default to full_band
        configuration = "full_band";
      }

      // Remove liberation marker from notes
      notes = notes.replace(/-=LIBERATED on [^=]+=- ?/g, "").trim();
      if (notes === "") notes = undefined;

      entries.push({
        songId,
        songTitle,
        lastPlayedDate,
        lastPlayedShowId,
        daysSince,
        showsSince,
        notes,
        configuration,
        isLiberated,
        liberatedDate,
        liberatedShowId,
      });
    });

    console.log(`Parsed ${entries.length} liberation list entries`);
    return entries;
  } catch (error) {
    console.error(`Error parsing Liberation page:`, error);
    return [];
  }
}

// Main scraper function
export async function scrapeLiberationList(): Promise<ScrapedLiberationEntry[]> {
  console.log("Starting liberation list scraper...");

  const browser = await chromium.launch({ headless: true });
  const page = await browser.newPage();

  // Set user agent to be polite
  await page.setExtraHTTPHeaders({
    "User-Agent":
      "DMBAlmanacClone/1.0 (Educational Project; Respectful Scraping)",
  });

  try {
    const entries = await parseLiberationPage(page);
    return entries;
  } finally {
    await browser.close();
  }
}

// Save liberation list to JSON file
export function saveLiberationList(entries: ScrapedLiberationEntry[]): void {
  const output: LiberationOutput = {
    scrapedAt: new Date().toISOString(),
    source: `${BASE_URL}/Liberation.aspx`,
    totalItems: entries.length,
    entries,
  };

  const filepath = join(OUTPUT_DIR, "liberation.json");
  writeFileSync(filepath, JSON.stringify(output, null, 2), "utf-8");
  console.log(`Saved ${entries.length} liberation list entries to ${filepath}`);
}

// Run if called directly
if (import.meta.url === `file://${process.argv[1]}`) {
  scrapeLiberationList()
    .then((entries) => {
      saveLiberationList(entries);
      console.log("Done!");
    })
    .catch((error) => {
      console.error("Scraper failed:", error);
      process.exit(1);
    });
}

File: projects/dmb-almanac/rust/crates/dmb_pipeline/src/scrape/liberation.rs
use anyhow::Result;
use scraper::Html;
use serde_json::{Value, json};

use super::{
    BASE_URL, ScrapeClient, normalize_whitespace, parse_dot_date, parse_i64_or_warn, regex,
    selector_or_warn, warn_if_empty, warn_missing_field,
};

pub(super) fn scrape_liberation(client: &ScrapeClient) -> Result<Vec<Value>> {
    let url = format!("{BASE_URL}/Liberation.aspx");
    let html = client.fetch_html(&url)?;
    Ok(parse_liberation_page(&html))
}

fn parse_liberation_page(html: &str) -> Vec<Value> {
    let document = Html::parse_document(html);
    let Some(row_selector) = selector_or_warn(
        "table.liberation-list tbody tr",
        "table.liberation-list tbody tr",
    ) else {
        return Vec::new();
    };
    let Some(song_selector) = selector_or_warn(
        "td.rowcell.lj a[href*='summary.aspx']",
        "td.rowcell.lj a[href*='summary.aspx']",
    ) else {
        return Vec::new();
    };
    let Some(last_played_selector) = selector_or_warn(
        "td.rowcell.cj a[href*='TourShowSet']",
        "td.rowcell.cj a[href*='TourShowSet']",
    ) else {
        return Vec::new();
    };
    let Some(days_selector) = selector_or_warn(
        "td.rowcell.cj.d-none.d-sm-table-cell",
        "td.rowcell.cj.d-none.d-sm-table-cell",
    ) else {
        return Vec::new();
    };
    let notes_selector = selector_or_warn("td.endcell.lj", "td.endcell.lj");
    warn_if_empty(&document, &row_selector, "liberation", "row");
    warn_if_empty(&document, &song_selector, "liberation", "song_link");
    warn_if_empty(
        &document,
        &last_played_selector,
        "liberation",
        "last_played_link",
    );
    warn_if_empty(&document, &days_selector, "liberation", "days_cells");

    let mut entries = Vec::new();
    for row in document.select(&row_selector) {
        let is_liberated = row
            .value()
            .attr("style")
            .is_some_and(|v| v.contains("background-color"));
        let song_link = row.select(&song_selector).next();
        let last_played = row.select(&last_played_selector).next();
        if song_link.is_none() || last_played.is_none() {
            continue;
        }
        let Some(song_link) = song_link else {
            continue;
        };
        let song_title = normalize_whitespace(&song_link.text().collect::<String>());
        let song_id = song_link
            .value()
            .attr("href")
            .and_then(|href| regex(r"sid=(\\d+)").captures(href))
            .and_then(|cap| cap.get(1).map(|m| m.as_str().to_string()))
            .unwrap_or_default();
        if song_id.is_empty() {
            warn_missing_field("liberation", "songId");
        }

        let Some(last_played_link) = last_played else {
            continue;
        };
        let last_played_date_text =
            normalize_whitespace(&last_played_link.text().collect::<String>());
        let last_played_date = parse_dot_date(&last_played_date_text).unwrap_or_else(|| {
            warn_missing_field("liberation", "lastPlayedDate");
            String::new()
        });
        let last_played_show_id = last_played_link
            .value()
            .attr("href")
            .and_then(|href| regex(r"id=(\\d+)").captures(href))
            .and_then(|cap| cap.get(1).map(|m| m.as_str().to_string()))
            .unwrap_or_default();
        if last_played_show_id.is_empty() {
            warn_missing_field("liberation", "lastPlayedShowId");
        }

        let mut days_since = 0;
        let mut shows_since = 0;
        let mut cells = row.select(&days_selector);
        if let Some(cell) = cells.next() {
            let text = normalize_whitespace(&cell.text().collect::<String>());
            days_since = parse_i64_or_warn(Some(&text), "liberation", "daysSince");
        }
        if let Some(cell) = cells.next() {
            let text = normalize_whitespace(&cell.text().collect::<String>());
            shows_since = parse_i64_or_warn(Some(&text), "liberation", "showsSince");
        }

        let notes_cell = notes_selector
            .as_ref()
            .and_then(|selector| row.select(selector).next());
        let mut notes = notes_cell
            .as_ref()
            .map(|cell| normalize_whitespace(&cell.text().collect::<String>()))
            .unwrap_or_default();
        let mut liberated_date = None;
        let mut liberated_show_id = None;
        if let Some(cell) = notes_cell
            && is_liberated
            && let Some(link) = cell.select(&last_played_selector).last()
        {
            if let Some(href) = link.value().attr("href") {
                liberated_show_id = regex(r"id=(\\d+)")
                    .captures(href)
                    .and_then(|cap| cap.get(1).map(|m| m.as_str().to_string()));
            }
            let link_text = normalize_whitespace(&link.text().collect::<String>());
            liberated_date = parse_dot_date(&link_text);
        }
        if is_liberated && liberated_show_id.as_deref().unwrap_or("").is_empty() {
            warn_missing_field("liberation", "liberatedShowId");
        }

        let notes_lower = notes.to_lowercase();
        let configuration =
            if notes_lower.contains("dave & tim") || notes_lower.contains("dave and tim") {
                "dave_tim"
            } else if notes_lower.contains("dave solo") {
                "dave_solo"
            } else {
                "full_band"
            };
        if notes.contains("LIBERATED") {
            notes = notes.replace("-=LIBERATED", "").trim().to_string();
        }

        entries.push(json!({
            "songId": song_id,
            "songTitle": song_title,
            "lastPlayedDate": last_played_date,
            "lastPlayedShowId": last_played_show_id,
            "daysSince": days_since,
            "showsSince": shows_since,
            "notes": if notes.is_empty() { Value::Null } else { Value::String(notes) },
            "configuration": configuration,
            "isLiberated": is_liberated,
            "liberatedDate": liberated_date,
            "liberatedShowId": liberated_show_id,
        }));
    }

    if entries.is_empty() {
        warn_missing_field("liberation", "entries");
    }
    entries
}
